User map

aapeliv opened the pull request:
Introduces some proxy trickery with PostGIS functions and a mapbox-gl based clustered user map

aapeliv committed:
Get basic gRPC transcoding working

aapeliv committed:
Output a basic geojson of users

aapeliv committed:
Implement ugly map of users

aapeliv committed:
Cluster points on map

aapeliv committed:
Clean up

aapeliv commented:
Can possibly be done in sqlalchemy too with something like this:

```py
res = session.query(func.ST_AsGeoJSON(tuple_(User.name, User.username, User.geom))).all()
```

kalvdans commented:
Why suddenly json when we have protobuf?

aapeliv commented:
Well there's two parts.

Why not gRPC?: because it's a bit tricky to hack these mapping libraries to load data in custom ways from custom sources (i.e. not from a HTTP endpoint). It's possible with a lot of fighting, but would then require hacks for each library/platform. It's also especially annoying because once you move to bigger GIS datasets, you have to start tiling it, which requires getting deeper into the mapping library's tile loading systems to hook in our gRPC transport.

Why GeoJSON: it's industry standard and is easy to use (+ has ready postgis functions). I'm not sure if you can do tiling with GeoJSON, at some point we'll probably want to move to some of the other formats, e.g. MVT.

kalvdans commented:
Sounds like splitting off user maps into a web-only service is better; avoids the gRPC mess at all. Non-web client apps can do fine without the user maps.

aapeliv commented:
> Sounds like splitting off user maps into a web-only service is better; avoids the gRPC mess at all.

Then we need to reimplement authentication for that service?

> Non-web client apps can do fine without the user maps.

Surely we want maps on mobile apps too?

aapeliv committed:
Update to not use session factories

lucaslcode commented:
Are we randomising points within the radius or displaying the circle with radius as given?

aapeliv commented:
> Are we randomising points within the radius or displaying the circle with radius as given?

When users choose the circle (center + radius) in the frontend, we should show that. At the moment this just draws a point at the center, and ignores the radius (which is fine when zoomed out, but could be fixed when zoomed in). 

I'm also not sure whether it really matters in a global map: we can show just the centers here, and when you choose a person, it can show the radius as well.

aapeliv commented:
As discussed in yesterday's backend meeting, this feature should not be merged yet before we think about the privacy implications of searching people by location, and give people an option to control what data is displayed.

amdwit commented on app/vue/src/api.ts:
```diff
@@ -1,4 +1,8 @@
-const URL = process.env.VUE_APP_API_URL
+export const URL = process.env.VUE_APP_API_URL
+
+export const ACCESS_TOKEN = process.env.VUE_APP_ACCESS_TOKEN
+
+export const nominatimURL = process.env.VUE_APP_NOMINATIM_URL
```
How important is pep 8? I see different naming styles here. Maybe we should use a style guide checker?
```suggestion
export const ACCESS_TOKEN = process.env.VUE_APP_ACCESS_TOKEN

export const NOMINATIM_URL = process.env.VUE_APP_NOMINATIM_URL
```

amdwit replied:
Now I see it is front-end code. Typescript. So no PEP 8. But maybe the google coding style.

aapeliv replied:
Yes, this is good to keep consistent. I will update it.

aapeliv commented:
I'd like to get this merged in as it has a lot of good stuff around shuffling non gRPC requests around. But we need to sort out the location privacy issue. I'll discuss it with Itsi.

lucaslcode commented:
Probably we at least need a "show me on the map" option

aapeliv committed:
Review comments

aapeliv committed:
Remove ErrorAlert that caused empty element at top of map

aapeliv committed:
Add protos.db to .gitignore

aapeliv committed:
Make GIS servicer secure

aapeliv committed:
Send credentials with GeoJSON requests from map

aapeliv committed:
Add extremely simple push to user page upon clicking user on map

aapeliv committed:
Fix typescript stuff

aapeliv committed:
Add zoom controls to maps

aapeliv committed:
Add map hash to url

aapeliv committed:
Write about location

aapeliv committed:
Tweak location selection component

Move buttons to top.

Hide searched address and only show randomized address.

Zoom out a bit by default.

aapeliv commented:
I've now also tweaked the map, so this PR closes #408 as well.

aapeliv commented:
I don't know how the "closes" parsing works.

Closes #408.

kalvdans commented on app/pb/google/api/httpbody.proto:
```diff
@@ -0,0 +1,77 @@
+
+import "google/protobuf/any.proto";
+
+option cc_enable_arenas = true;
+option go_package = "google.golang.org/genproto/googleapis/api/httpbody;httpbody";
+option java_multiple_files = true;
+option java_outer_classname = "HttpBodyProto";
+option java_package = "com.google.api";
+option objc_class_prefix = "GAPI";
```
Please keep our protocol definition language-agnostic.

Would it be possible to draw in these files via pip instead of copying them?

I feel reluctant to add third party files.

aapeliv replied:
Indeed, these are copied directly from the google-api definitions [here](https://github.com/googleapis/googleapis). There is a python package that provides these [here](https://pypi.org/project/googleapis-common-protos/) ([github](https://github.com/googleapis/api-common-protos)), but it seems out of date/out of sync with the googleapis repo. Also, I believe we need the full `.proto` definitions in order to get the full descriptor set for envoy, so we would need to get the protos into the `protoc` include path.

For me, it seemed that this sin was justified and much simpler than trying to get the dependencies easily installable across platforms. It's all standalone protobufs anyway, I don't see envoy changing their transcoding definitions any time soon.

We could look for another option, but I don't think it's worth the hassle at the moment, there's a lot of other important things to work on!

(These lines don't restrict language compatibility, they just define some extra options for how the code should be generated for certain languages.)

(Also, small note: I've modified the `include`s here to include from our `pb/` folder.)

kalvdans commented on app/backend/src/couchers/servicers/gis.py:
```diff
@@ -0,0 +1,29 @@
+
+            return httpbody_pb2.HttpBody(
+                content_type="application/json",
+                data=json.dumps(out.scalar()).encode("utf8"),
```
`json.dumps` only generate ascii by default, so use `.encode("ascii")`.

kalvdans commented on app/backend/src/couchers/servicers/gis.py:
```diff
@@ -0,0 +1,29 @@
+import grpc
+
+from couchers.db import session_scope
+from pb import gis_pb2, gis_pb2_grpc
```
`gis_pb2` is unused.

kalvdans reviewed (dismissed):
I've only reviewed the backend code, and not tested it in any way. Just some small remarks. Ok to merge for me with or without fixing them :)

aapeliv committed:
Review comments

aapeliv commented:
Thanks @kalvdans: fixed!

lucaslcode reviewed (approved):
LGTM but didn't test. No CI on this branch? Was it opened before then?

aapeliv commented:
> No CI on this branch

They ran for me

https://gitlab.com/couchers/couchers/-/pipelines/232254459
